Add `ConvertTo-CliXml` and `ConvertFrom-CliXml` cmdlets

ArmaanMcleod opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->

Fixes #3898

Add `ConvertTo-CliXml` and `ConvertFrom-CliXml` cmdlets to work with CliXml objects in memory without needing to read/write to file system.

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

### Parameter sets

```pwsh
ConvertTo-CliXml [-InputObject] <psobject> [-Depth <int>] [<CommonParameters>]

ConvertFrom-CliXml [-InputObject] <string> [<CommonParameters>]
```

### Usage

```pwsh
> $cliXmlString = [pscustomobject]@{'hello' = 1} | ConvertTo-CliXML
> $cliXmlString
<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04">
  <Obj RefId="0">
    <TN RefId="0">
      <T>System.Management.Automation.PSCustomObject</T>
      <T>System.Object</T>
    </TN>
    <MS>
      <I32 N="hello">1</I32>
    </MS>
  </Obj>
</Objs>
> $cliXmlString | ConvertFrom-CliXml

hello
-----
    1
```

Alot of this was inspired from #12845, but that PR is quite old and never got merged. I just created a new PR and made some changes since it was easier. 

Also included `internal static string Serialize(IList<object> source, int depth, bool enumerate)` in `PSSerializer` API. This was to be able to enumerate and serialize objects one at a time instead of writing one top level object. 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/10773
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

ArmaanMcleod commented on test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:
```diff
@@ -1,5 +1,9 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
+
+[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingConvertToSecureStringWithPlainText', '')]
+param()
```
Was added to just suppress the codefactor issue in 26daabad5. Happy to remove if we don't want to suppress.

iSazonov commented:
Did you address https://github.com/PowerShell/PowerShell/pull/12845#issuecomment-891349031?

ArmaanMcleod commented:
> Did you address [#12845 (comment)](https://github.com/PowerShell/PowerShell/pull/12845#issuecomment-891349031)?

I'm not sure if I agree with that comment. The existing `Import-CliXml`/`Export-CliXml` cmdlets don't even use the PSSeralizer APIs, they use `InternalSerializer`/`InternalDeserializer` classes. 

It can probably be done for `ConvertFrom-CliXml`/`ConvertTo-CliXml` but there is a lot of rework since `PSSerializer` don't have access to `ImportXmlHelper` class and methods like `Import()`.

Also reasoning for using `PSSerializer` APIs doesn't make sense to me still, they are using .NET serializer API under the hood anyways.

iSazonov commented:
> Also reasoning for using `PSSerializer` APIs doesn't make sense to me still, they are using .NET serializer API under the hood anyways.

There are security "wrappers" specific to PowerShell. We should follow that.
As for PSSerializer it was mentioned as _public_ API. I think it is based on InternalSerializer/InternalDeserializer. So, if it is impossible use the public API you could use internal API. I suggest you directly consult with @TravisEz13 before you start the security sensitive work.

ArmaanMcleod commented:
> > Also reasoning for using `PSSerializer` APIs doesn't make sense to me still, they are using .NET serializer API under the hood anyways.
> 
> There are security "wrappers" specific to PowerShell. We should follow that. As for PSSerializer it was mentioned as _public_ API. I think it is based on InternalSerializer/InternalDeserializer. So, if it is impossible use the public API you could use internal API. I suggest you directly consult with @TravisEz13 before you start the security sensitive work.

I think for using security wrapper PSSerializer it is possible, but it requires creating more overloads for cmdlet `PagingParameters` to get similar implementation of `XmlHelper.Import()` code. That is currently not possible in `PSSerializer class alone. If that is something that MSFT team is fine with, we can do it, otherwise we have to use internal API code already being used for import/export CLIXML cmdlets.

ArmaanMcleod commented:
@iSazonov @TravisEz13 I have converted code to use `PSSerializer` API but had to include some overloads to support `PagingParameters`, which is basically a copy of `XmlHelper.Import()` implementation, but returns a list of objects instead. This was the only way to address security concern and use `PSSerializer` API.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -632,6 +632,94 @@ private void CleanUp()
+    /// Implements ConvertFrom-CliXml command.
+    /// </summary>
+    [Cmdlet(VerbsData.ConvertFrom, "CliXml", SupportsPaging = true, HelpUri = "")]
+    public sealed class ConvertFromClixmlCommand : PSCmdlet
```
@sdwheeler We need HelpUri-s for the new cmdlets.

iSazonov commented on test/powershell/engine/Basic/DefaultCommands.Tests.ps1:
```diff
@@ -240,6 +241,7 @@ Describe "Verify aliases and cmdlets" -Tags "CI" {
 "Cmdlet",       "ConvertFrom-StringData",           "",                                 $($FullCLR -or $CoreWindows -or $CoreUnix),     "",                     "",                     "None"
 "Cmdlet",       "Convert-Path",                     "",                                 $($FullCLR -or $CoreWindows -or $CoreUnix),     "",                     "",                     "None"
 "Cmdlet",       "Convert-String",                   "",                                 $($FullCLR                               ),     "",                     "",                     ""
+"Cmdlet",       "ConvertTo-CliXml",                 "",                                 $($FullCLR -or $CoreWindows -or $CoreUnix),     "",                     "",                     "None"
```
I wonder if the cmdlets were in Windows PowerShell.

ArmaanMcleod replied:
I dont believe so. They are new for all OS platforms.

iSazonov replied:
I mean we should remove `$FullCLR`

ArmaanMcleod replied:
Fixed in 0a72c4aee

ArmaanMcleod commented:
Made a new `PSSerializer.WriteDeserializedObjects` internal method here which is more close to the `XmlHelper.Import()` method as it writes deserialized objects to pipeline directly instead of creating list.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -632,6 +632,94 @@ private void CleanUp()
+        /// </summary>
+        protected override void ProcessRecord()
+        {
+            _inputObjectBuffer.Add(InputObject);
```
What if InputObject is array/list/collection? What is right behavior in the case? Have we tests for the scenario?

ArmaanMcleod replied:
I've added some tests in e3c4c4f3d. Behaviour seems to hold up but can add more as we find to get more coverage.

iSazonov replied:
My concern is about `$content = ConvertTo-CliXml -InputObject @(1,2,3,4)`

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -632,6 +632,94 @@ private void CleanUp()
+                // If single input object don't wrap in array
+                object source = _inputObjectBuffer.Count == 1
+                    ? _inputObjectBuffer[0]
+                    : _inputObjectBuffer.ToArray();
```
I guess this correlate with my previous comment.

ArmaanMcleod replied:
This is pretty much identical to how `ConverTo-Json` works, which forces singleton item array to be just one item returned.

mklement0 replied:
Just an aside: it's understandable to think the `ConvertFrom-Json` parameter name is `-AsArray` (I'm tempted to think that every time), but it's actually `-NoEnumerate`, and it doesn't _ensure_ that the output is an array; it only outputs an array (as a whole) if the conversion result happens to be an array. By contrast, `ConvertTo-Json`'s `-AsArray` currently _unconditionally constructs an array wrapper_, which I think is problematic with direct use of `-InputObject` - see #10952

iSazonov replied:
Then this code doesn't make sense here.

ArmaanMcleod replied:
Thanks @mklement0 for explanation and reminding me, I often get those parameters mixed up in my head 😄.

In this case not sure what behaviour we should opt in for this cmdlet. @iSazonov what do you suggest instead so this makes sense?

iSazonov replied:
We have no need to force output in an array. 
We should get the same result as Export-Clixml output.

mklement0 replied:
The code looks OK to me for the direct `-InputObject` argument case, but `Export-CliXml` actually works differently with _pipeline_  input (I haven't looked a how and why):

E.g., the following doesn't serialize as an array, but as two top-level integers:

```powershell
 1, 2 | Export-Clixml  temp.xml; Get-Content temp.xml
```

```
<Objs Version="...">
  <I32>1</I32>
  <I32>2</I32>
</Objs>
```

ArmaanMcleod replied:
I think this is quite tricky because `PSSerializer.Serialize` calls `WriteOneTopLevelObject`, so in this case all the input is collected into `EndProcessing` and it gets forced to an array. Will need to see how `Export-CliXml` does it, which btw is not using `PSSerializer` APIs 😄.

ArmaanMcleod replied:
The internal APIs that `Export-CliXml` use basically serialize each input object one at a time then writes the root element with `serialize.Done()` at the end. The `PSSerializer.Serialize` API however writes one whole top level object, so basically its not possible it seems to get the same behaviour as is.

ArmaanMcleod replied:
What I think will be needed is a new overload for `PSSerializer.Serialize` that serializes each input one at a time instead of everything at once like with existing code. Will try that later.

iSazonov replied:
Perhaps all these cmdlets should be based on the same api. to work identically. Then there are two ways. Just use the internal api now and switch to the public api later. Or make a public api now and migrate the rest of the cmdlets later.
Perhaps first way is more simple.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -632,6 +632,94 @@ private void CleanUp()
+        /// Gets or sets input object which is written in CliXml format.
+        /// </summary>
+        [Parameter(Position = 0, Mandatory = true, ValueFromPipeline = true)]
+        public string InputObject { get; set; }
```
Maybe PSObject?

ArmaanMcleod replied:
Wasn't sure about this one. To deserialize you need a string so didn't see why we'd keep it as PSObject.

iSazonov replied:
@mklement0 What do you think?

mklement0 replied:
@ArmaanMcleod, thanks for tackling this and the `Show-Markdown` issue. 
@iSazonov, using `string` makes sense to me, and it's also what `ConvertFrom-Json` and `ConvertFrom-StringData` do. Do you have specific concerns?

iSazonov replied:
@mklement0 I remember https://github.com/PowerShell/PowerShell/issues/4242 So it would be great to have you conclusion about the parameter and how the cmdlet process it.

mklement0 replied:
@iSazonov, by using `string` we avoid conceptual ambiguity and clearly communicate that the cmdlet can only accept _multiple_ input objects (strings) via the _pipeline_, as most of the existing cmdlets do. While for some cmdlets the ability to pass multiple objects via an array directly to `-InputObject` is important for performance reasons, I don't think we need to worry about it here: I expect the typical use case to be a _single_ input string. (Multiple input strings often come from lazy use of `Get-Content` rather than `Get-Content -Raw` as input, but the file-input case is covered by `Import-CliXml`).

iSazonov replied:
Wouldn't there be a problem with aggressive conversion to a string?

> Multiple input strings often come from lazy use of Get-Content

Original scenario was `Get-WhatsNew | Show-Markdown` where first cmdlet outputs an array of strings.

mklement0 replied:
@iSazonov, you're thinking of a specific `Show-Markdown` use case; I can't think of an analogous scenario for `ConvertFrom-CliXml`, whose input I'd typically expect to be a single string emitted by `ConvertTo-CliXml` (as noted, with _file_ input you'd just use `Import-CliXml`).

iSazonov replied:
Ah, sorry :-) There is another PR for Show-Markdown. My comment is in the wrong place.

mklement0 replied:
@iSazonov  👍, easy to get confused.
On an unrelated note: at least at first glance it looks like `Import-CliXml`'s `-IncludeCount` switch does nothing useful and always outputs a dummy result (a `[psobject]` that wraps `[double]` `0.0` and stringifies to `Unknown total count`). Do you know the backstory? If my analysis is correct, at least the docs should be updated.

iSazonov replied:
@mklement0 I don't know the history. I see Engine has formal support for "paging" parameters (like common and dynamic parameters). Only Import-CliXml has one line code to check for "paging" parameter - IncludeTotalCount - with zero values. This cmdlet definitely doesn't assign SupportsPaging to true in the CmdletAttribute and it's dead code. We could remove it.

ArmaanMcleod replied:
I was also thinking if it's even worth including `SupportsPaging` for these new CliXml cmdlets. I included it for consistency but honestly `IncludeCount` is totally useless right now, and `-First` and `-Skip` can be achieved with `Select-Object`. I'm not even convinced the performance is better since `SupportsPaging` still requires so much complicated code to filter. I could be totally wrong here since I havn't used `SupportsPaging` before, probably need say from others who have more background.

iSazonov replied:
Obviously, we have no need for SupportsPaging. Please remove.
(We could discuss the feature and its history in new issue if you have an interest.)

ArmaanMcleod replied:
Done

ArmaanMcleod commented:
> @iSazonov @TravisEz13 I have converted code to use `PSSerializer` API but had to include some overloads to support `PagingParameters`, which is basically a copy of `XmlHelper.Import()` implementation, but returns a list of objects instead. This was the only way to address security concern and use `PSSerializer` API.
> 
> If introducing these overloads is not acceptable, I can go back to previous approach with internal API. I can also not make the access modifiers not `public` and maybe do `internal` as well if that is better.

Reverted this as we don't see a new for `SupportsPaging` capability in this cmdlet, since `-IncludeTotalCount` is not helpful and `-First`/`-Skip` can be achieved with `Select-Object`.

ArmaanMcleod commented:
Thanks @mklement0 and @iSazonov for comprehensive and very helpful reviews 🙂.

I have included an `PSSerializer` overload for the time-being `internal static string Serialize(object[] source, int depth, bool enumerate)`, and seems to work well and matches up behaviour with Import/Export cmldets. Happy to change this but I think this is an easy way to solve problem. I've kept it `internal` since don't want to expose this API unless its ready to go public, since I think we will go through more iterations in this PR and when we refactor the other CliXML cmdlets to use public API. I think there are many ways to do the API but main thing is we need the enumerate logic for serialization to avoid writing one whole top level object each time.

Please review if you get a chance and let me know if I've missed anything.

ArmaanMcleod commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -122,6 +122,45 @@ public static string Serialize(object source, int depth)
+        /// <param name="depth">The depth of the members to serialize.</param>
+        /// <param name="enumerate">Enumerates input objects and serializes one at a time.</param>
+        /// <returns>The serialized object, as CliXml.</returns>
+        internal static string Serialize(object[] source, int depth, bool enumerate)
```
Fixed in 06aafe66a

ArmaanMcleod commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -632,6 +632,91 @@ private void CleanUp()
+        /// </summary>
+        [Parameter]
+        [ValidateRange(1, int.MaxValue)]
+        public int Depth { get; set; } = 1;
```
This was interesting. I wonder how it worked in `Export-CliXml` since `ValidateRange` should error on default of 0?

mklement0 replied:
I haven't looked at the code, but I found in my notes the following snippet, implying that `Export-CliXml` uses `2` as the depth.

```powershell
class Bar {
  [int] $Baz = 42
}
class Foo {
  [Bar] $Baz = [Bar]::new()
}

$o = [Foo]::new()

[System.Management.Automation.PSSerializer]::Serialize($o)

Write-Host '---'

$o | Export-Clixml t.xml

Get-Content t.xml
```

iSazonov replied:
@ArmaanMcleod Could you please re-check? I believe it is more safe use explicit default here and in other cmdlets.

ArmaanMcleod replied:
So Export-CliXml doesn't set an explicit default for the parameter,

```pwsh
# Export-CliXml
[Parameter]
[ValidateRange(1, int.MaxValue)]
public int Depth { get; set; }
```

I wonder if it's doing this in the background somewhere, couldn't see in the code where its being done, or if internal API serializer has default depth of 2 somewhere.

But I guess in this case if it observed to have default depth of 2 then I can do the same for `ConvertTo-CliXml`. Might just be easier to be explicit and set a default to 2 to the parameter directly.

There is some interesting code for `Export-CliXml` which checks Depth == 0, which is for when depth is not specified since 0 is default for `int`, but I think in the background the depth is really 2 from default. 

Docs seem to also indicate default depth is 2: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/export-clixml?view=powershell-7.4#-depth

iSazonov replied:
> couldn't see in the code where its being done, or if internal API serializer has default depth of 2 somewhere.

Is the default in .Net?
-----

Have we tests for the default for both cmdlets? If no, please add to avoid regressions in future.

ArmaanMcleod replied:
Could not find in .Net API but I've added tests for default of 2. Should be ok for now to avoid regression.

mklement0 commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -122,6 +122,45 @@ public static string Serialize(object source, int depth)
+            XmlWriterSettings xmlSettings = new()
+            {
+                CloseOutput = true,
+                Encoding = Encoding.Unicode,
```
Given that the cmdlet outputs a `string` instance, I don't think character encoding comes into play here (it only would when saving to a file or transmitting via the network).

ArmaanMcleod replied:
Thanks @mklement0. Yeah I thought the same and didn't see value in encoding strings in memory, only useful for files.

iSazonov replied:
+1 to remove

iSazonov commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -122,6 +122,45 @@ public static string Serialize(object source, int depth)
+            XmlWriter xw = XmlWriter.Create(sb, xmlSettings);
+            Serializer serializer = new(xw, depth, useDepthFromTypes: true);
+
+            if (enumerate)
```
How do this work if parameter is IEnumerable and enumerate is false?

ArmaanMcleod replied:
It will just serialize the entire enumerable at once. Do you think we should set a default for this? I was thinking for public API eventually you van configure this behaviour to enumerate or write entire enumerable at once.

ArmaanMcleod replied:
Basically if enumerate is false we get behaviour from the other public Serialize method.

iSazonov replied:
The question is, what if the argument is IEnumerable and enumerate = false? It doesn't seem to work.

ArmaanMcleod replied:
Interesting. Wonder if Serializer has issues with Ienumerable. Will need to write some Xunit tests to verify.

ArmaanMcleod replied:
@iSazonov I've changed to `IList<object> source`. Seems to have better behaviour. Added XUnit tests for both cases and it seems good, let me know what you think 🙂

iSazonov replied:
Formally LGTM.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -340,4 +346,181 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+    }
+
+    It "ConvertTo-CliXml should serialize PSCustomObject correctly" {
+        $testObject = [PSCustomObject]@{ a = 1; b = 2; c = 3; d = 4 }
```
Maybe use one nested object?

ArmaanMcleod replied:
Added test for nested object in d90b55133

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -340,4 +346,195 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
 		$cmd.Xml | Should -Be $xml
 	}
+
+	It "ConvertFrom-CliXml directive should work" {
+        $content = Get-Command export* -Type Cmdlet | Select-Object -First 3 | ConvertTo-CliXml
+        $results = ConvertFrom-CliXml $content
+        $results.Count | Should -Be 3
+        $results[0].PSTypeNames[0] | Should -BeExactly "Deserialized.System.Management.Automation.CmdletInfo"
+    }
```
```suggestion
	It "ConvertFrom-CliXml directive should work" {
            $content = Get-Command export* -Type Cmdlet | Select-Object -First 3 | ConvertTo-CliXml
            $results = ConvertFrom-CliXml $content
            $results.Count | Should -Be 3
            $results[0].PSTypeNames[0] | Should -BeExactly "Deserialized.System.Management.Automation.CmdletInfo"
    }
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -340,4 +346,195 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+        $out[2] | Should -Be 3
+    }
+
+    It "ConvertTo-CliXml using default depth of 2 should work" {
+        $testObject = [One]::New()
+		$content = $testObject | ConvertTo-CliXml
+		$testObject | Export-Clixml -Path $testfile
+		$content | Should -Be (Get-Content -Path $testfile -Raw)
+		$deserialized_one = ConvertFrom-CliXml -InputObject $content
+		$deserialized_one.value | Should -Be 1
+        $deserialized_one.two | Should -Not -BeNullOrEmpty
+		$deserialized_one.two.value | Should -Be 2
+		$deserialized_one.two.three | Should -Not -BeNullOrEmpty
+		$deserialized_one.two.three.num | Should -BeNullOrEmpty
+    }
```
```suggestion
    It "ConvertTo-CliXml using default depth of 2 should work" {
        $testObject = [One]::New()
        $content = $testObject | ConvertTo-CliXml
        $testObject | Export-Clixml -Path $testfile
        $content | Should -Be (Get-Content -Path $testfile -Raw)
        $deserialized_one = ConvertFrom-CliXml -InputObject $content
        $deserialized_one.value | Should -Be 1
        $deserialized_one.two | Should -Not -BeNullOrEmpty
        $deserialized_one.two.value | Should -Be 2
        $deserialized_one.two.three | Should -Not -BeNullOrEmpty
        $deserialized_one.two.three.num | Should -BeNullOrEmpty
    }
```

iSazonov reviewed (commented):
@ArmaanMcleod Please replace tabs in tests if you see them.

And if you don't plan new commits I suggest squash commits (to many commits, too many comments) and open new PR.

ArmaanMcleod committed:
Add ConvertFrom-CliXml & ConvertTo-CliXml cmdlets

ArmaanMcleod committed:
Add tests

ArmaanMcleod committed:
Fix new overload XML doc summary

ArmaanMcleod committed:
Grouping cmdlets under context and simplifying test titles

ArmaanMcleod commented:
> @ArmaanMcleod Please replace tabs in tests if you see them.
> 
> And if you don't plan new commits I suggest squash commits (to many commits, too many comments) and open new PR.

Thanks @iSazonov 

I have opened up a new PR #21011. Closing this one in favour of that one.
